[Doc] ADR-01 — core architecture and community decisions (#13)
vllm-project/agentic-stack: docs/adr/ADR-01_core.md

# ADR-01 — Project Core
> **Status:** Draft — open for community review
> **Language:** Python (Red Hat, DaoCloud, EmbeddedLLM voted Python; Meta leaned Rust but deferred)
> **Deciders:** Francisco Arceo, Sébastien Han, Ben Browning, Andrew Xia, Tun Jian Tan, Jia Huei Tan, Cheong Ye Hur, Maral Bahari

---

## 1. Overview

> This ADR covers the foundational architecture of **agentic-stack** — an open-source gateway that sits in front of a stateless LLM inference backend and turns it into a stateful, tool-capable Responses API endpoint.

Reviewer: yes

Three core challenges:

> 1. **Statefulness.** Clients shouldn't have to send their full conversation history every request. The Responses API lets the client reference a prior response by ID and send only new input. The gateway needs to bridge that gap.

Reviewer 2: ```suggestion
1. **Statefulness.** Clients shouldn't have to send their full conversation history or `prompt_id` in every request. The Responses API lets the client reference a `previous_response_id` and `prompt_id` and send only the new input. The project needs to bridge that gap.
```

| 1 | **Entry Points** | Process startup, signal handling, ASGI server integration. Should be thin wrappers that assemble the app and hand off to HTTP. |
| 2 | **HTTP Routing** | Routes, request parsing/validation, response serialization. Produces structured objects for the orchestration layer. Owns the SSE streaming loop. |
| 3 | **Core Orchestration** | The main request lifecycle — history rehydration, upstream LLM calls, protocol translation, tool dispatch, response persistence. Most of the complex logic lives here. |
| 4 | **Tool Runtimes** | Built-in tool implementations (code interpreter, web search). Self-contained with a clear interface: start, health-check, execute, shut down. |
| 5 | **MCP Integration** | MCP server management (hosted + request-remote), tool name mapping, security enforcement. |
| 6 | **Config & Shared Foundations** | Configuration, DB connections, observability, shared utilities. Should not depend on layers 1-5. |

For the MVP, focus is on layers 1-3 and 6. Community discussion was clear about starting minimal — single-device deployment, works out of the box, no config needed. Tool Runtimes and MCP Integration come in a later phase.

> On the response store specifically: we're leaning toward a single-table design where each row holds a JSON blob with the full flattened conversation history. That way rehydrating a conversation for `previous_response_id` is always one DB read — no walking a chain of linked responses. SQLite as the default backend (zero-config), PostgreSQL when you need multi-worker, and an optional Redis layer for hot lookups. We'll flesh this out properly in a follow-up ADR once the core architecture here settles.

Reviewer 2: We can have some generic classes that enable database providers to choose whatever is best fit for them (e.g., kv or entity lookup). 

Agree on sqlite and we can probably have a configuration to allow users to change defaults.
